### DISCUSSION TITLE
A way to add buttons to the form

### DISCUSSION THREAD

**Discussion by yob-yob** [filamentphp/filament#285] [Ideas]
now the create-form page has the button `save`.

as a user

I want to have something like `Cancel`,  `Save and close`, `Save and create another` & `Save` buttons
and these buttons can be turned off since I want to turn them on some resources only.

so that it's simple for the customers to understand what's happening when they clicked a button.

**Comment by yob-yob** [comment]
or maybe much better

as a user

I want to be able to add buttons to `create-form` page and `edit-form` page.

so that it's easy for me to add any actions I want need to make the form easy to use.

an example would be to add `save and close` functionality

**Reply by Whizboy-Arnold** [reply]
Or just abstract buttons on a layout anywhere on the page or at header section for instance and link to abstract action

**Comment by dev3k** [comment]
Not sure if you found solution for this yet, but i needed such button and i was able to do it like below

in your edit resource page add the following

```
protected function getFormActions(): array
    {
        return [
            $this->getSaveFormAction(),
            Action::make('saveAnother')
                ->label('Save & create another')
                ->action('saveAnother')
                ->keyBindings(['mod+shift+s'])
                ->color('secondary'),
            $this->getCancelFormAction(),
        ];
    }

public function saveAnother()
    {
        $resources = static::getResource();
        $this->redirect($resources::getUrl('create'));
        $this->save(true);
    }
```

**Reply by Whizboy-Arnold** [reply]
sorry doesn't work. Also am on on V2

**Reply by dev3k** [reply]
I tested it using the most recent version, v2.17.38, and it appears to be functioning properly. However, I did observe that it redirects to new page even when validation fails, which could have been the issue you experienced, @Whizboy-Arnold 


```php
public function saveAnother()
    {
        $resources = static::getResource();
         $this->save(true); // I moved this up , so now it triggers validation and stop the redirect if validation fails
        $this->redirect($resources::getUrl('create'));
    }
```

**Reply by Whizboy-Arnold** [reply]
Yeap, that was the problem. 
Also however these actions are only available on form page. not the other pages I would prefer an abstract way to just add buttons on a page @dev3k

**Reply by jeannegoossens** [reply]
Updated it for V3:

```php
protected function getFormActions(): array
    {
        return [
            $this->getCreateFormAction(),
            Action::make('saveAnother')
                ->label('Save & create another')
                ->action('saveAnother')
                ->keyBindings(['mod+shift+s'])
                ->color('secondary'),
            $this->getCancelFormAction(),
        ];
    }

public function saveAnother()
    {
        $resources = static::getResource();
        $this->create();
        $this->redirect($resources::getUrl('create'));
    }
```

**Reply by dev3k** [reply]
@jeannegoossens  this code for create form which already has "create & create another" action , the edit page was and still "v3" missing this action, here is the code for edit form

```php
 protected function getFormActions(): array
    {
        return [
            $this->getSaveFormAction(),
            Actions\Action::make('saveAnother')
                ->label('Save & create another')
                ->action('saveAnother')
                ->keyBindings(['mod+shift+s'])
                ->color('gray'),
            $this->getCancelFormAction(),
        ];
    }

    public function saveAnother():void
    {
        $resources = static::getResource();
        $this->save(false);
        $this->redirect($resources::getUrl('create'));
    }
```

I think they removed the secondary color in v3, its called gray now?

**Comment by judgej** [comment]
Are you looking to just add a filament button to a page? This is [documented in V3](https://filamentphp.com/docs/3.x/support/blade-components/button), but just found it works on V2:

     <x-filament::button>My Button</x-filament::button>

It does not need to be on a form, table or admin component - it works anywhere. Maybe it is in V2 document? I've never been able to find it though.

**Reply by yob-yob** [reply]
this was literally 2 years ago...

**Reply by danharrin** [reply]
@yob-yob Helpful people comment recent solutions to help new readers out. No need to react like that.

**Reply by yob-yob** [reply]
Sorry about that.

**Reply by judgej** [reply]
I see posts and issues all as resources that stay alive as long as people have a problem to solve. This issue kept coming up when I was looking to add a general "filament button", so guessed it would be the same for others. I may have a completely wrong handle on what your original problem was, and will be happy to move it to the hints and tips section. The V3 documentation filling in gaps for the V2 documentation is a pretty cool though.

**Reply by commongoodguy** [reply]
very helpful, somehow this page is up top on google when I was searching for a way to have consistent styling when using Form Builder with Admin Panel custom page

**Reply by mattmintun** [reply]
helped me and here we are in 2024! Thank you @judgej

**Reply by a89s** [reply]
@mattmintun same :D

**Comment by alexmanase** [comment]
https://laraveldaily.com/post/filament-add-custom-action-button-inside-form

Maybe this can help you.

**Reply by dev3k** [reply]
nice, I didn't know you can add action button inside the form, thanks :)

**Comment by pabbec** [comment]
Hi!

I was able to add two buttons following these examples, but I need to modify the value (state) of a field if the user clicks on that button. Specifically:

I have two buttons:

Save to draft
Save and Publish

In both casds, I want to save the form. But, if the user clicks the "Save to draft" button, I need change the status field to the value "draft". And if the user clicks the "Save and publish" button, I need to set the status field to the value "published".

In my PostResource/Pages/CreatePost.php I have:

```
     protected function getFormActions(): array
     {
         return [
             $this->getCreateInDraftFormAction(),
             $this->getCreateAndConfirmFormAction(),
             $this->getCancelFormAction(),
         ];
     }

     protected function getCreateInDraftFormAction(): Action
     {
         return Action::make('createindraft')
         ->label('Save to Draft')
         ->submit('createindraft');
     }



     protected function getCreateAndConfirmFormAction(): Action
     {
         return Action::make('createandconfirm')
         ->label('Save and Confirm')
         ->submit('createandconfirm');
     }
```

But I can't find a way to access that field in my form and set the value I want.

**Reply by gediminas1994** [reply]
Hi,

I've encountered a similar problem and posted it on discord. Here's a reply I got from @leandrocfe and it worked like a charm!

```
Action::make('createindraft')
    ->label(Save to Draft')
    ->action(function(){
        $this->data['status'] = 'draft';
        $this->create();
    })
```

**Comment by raphaelcangucu** [comment]
For Filament V2 you can use something like this:

```
<x-filament::card>
    <form wire:submit.prevent="firstFormSubmit" class="space-y-6">

        {{ $this->firstForm}}

        <div class="text-right">
            <x-filament::button type="submit" wire:submit="firstFormSubmit" class="align-right">
               Submit first form
            </x-filament::button>
        </div>
    </form>
</x-filament::card>

<x-filament::card>
    <form wire:submit.prevent="otherFormSubmit" class="space-y-6">

        {{ $this->otherForm}}

        <div class="text-right">
            <x-filament::button type="submit" wire:submit="otherFormSubmit" class="align-right">
                Submit other form
            </x-filament::button>
        </div>
    </form>
</x-filament::card>
```

And in the Page you will need to forms:

```php
 protected function getForms(): array
    {
        return [
            'firstForm' => $this->makeForm()
                ->schema($this->getFirstFormSchema()),
 'otherForm' => $this->makeForm()
                ->schema($this->getOtherFormSchema()),

        ];
    }
```
